pkg/jobs/jobsauth/authorization.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package jobsauth
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/roleoption"
+)
+
+// An AccessLevel is used to indicate how strict an authorization check should
+// be.
+type AccessLevel int
+
+const (
+	// ViewAccess is used to perform authorization for viewing jobs (ex. SHOW JOBS).
+	ViewAccess AccessLevel = iota
+
+	// ControlAccess is used to perform authorization for modifying jobs (ex. PAUSE|CANCEL|RESUME JOB).
+	// This access level performs stricter checks than ViewAccess.
+	//
+	// The set of jobs visible via ControlAccess is a subset of jobs visible via
+	// ViewAccess. In other words: if a user with a given set of privileges is
+	// authorized to modify a job using ControlAccess, they will be authorized to
+	// view it using ViewAccess.
+	ControlAccess
+)
+
+var authorizers = make(map[jobspb.Type]Authorizer)
+
+// Authorizer is a function which returns a pgcode.InsufficientPrivilege error if
+// authorization for the job denoted by jobID and payload fails.
+type Authorizer func(
+	ctx context.Context, a AuthorizationAccessor, jobID jobspb.JobID, payload *jobspb.Payload,
+) error
+
+// RegisterAuthorizer registers a AuthorizationCheck for a certain job type.
+func RegisterAuthorizer(typ jobspb.Type, fn Authorizer) {
+	authorizers[typ] = fn
```

Comment:
Sorry, you should definitely panic if typ already exists in authorizers.

Reply:
Yeah I thought so. Will update.

Interestingly [register resumer](https://github.com/cockroachdb/cockroach/blob/bfdd3b0c61d212861702f6575f743b4b22be98f9/pkg/jobs/registry.go#L1387) doesn't panic. There doesn't seem to be a particular reason for that.